x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -191,51 +201,185 @@ public void authorize(final Authentication authentication, final String action,
             // the originating action is the current action if no originating action has yet been set in the current thread context
             // if there is already an original action, that stays put (eg. the current action is a child action)
             putTransientIfNonExisting(ORIGINATING_ACTION_KEY, action);
+            authorize(authentication, action, originalRequest, listener, this::authorizeStandaloneAction);
+        }
+    }
 
-            String auditId = AuditUtil.extractRequestId(threadContext);
-            if (auditId == null) {
-                // We would like to assert that there is an existing request-id, but if this is a system action, then that might not be
-                // true because the request-id is generated during authentication
-                if (isInternal(authentication.getUser())) {
-                    auditId = AuditUtil.getOrGenerateRequestId(threadContext);
-                } else {
-                    auditTrailService.get().tamperedRequest(null, authentication, action, originalRequest);
-                    final String message = "Attempt to authorize action [" + action + "] for [" + authentication.getUser().principal()
-                            + "] without an existing request-id";
-                    assert false : message;
-                    listener.onFailure(new ElasticsearchSecurityException(message));
-                    return;
-                }
-            }
+    private void authorize(final Authentication authentication, final String action, final TransportRequest originalRequest,
+                           final ActionListener<Void> listener,
+                           final TriConsumer<RequestInfo, String, ActionListener<Void>> actionAuthorization) {
+        String auditId;
+        try {
+            auditId = requireAuditId(authentication, action, originalRequest);
+        } catch (ElasticsearchSecurityException e) {
+            listener.onFailure(e);
+            return;
+        }
+
+        // sometimes a request might be wrapped within another, which is the case for proxied
+        // requests and concrete shard requests
+        final TransportRequest unwrappedRequest = maybeUnwrapRequest(authentication, originalRequest, action, auditId);
```

Comment:
Should we unwrap the request before checking preauthz for child actions? I think in theory it makes more sense to do so. But it may not have actual impact in practice?

Reply:
We should. It's on my list.